test/test_broker.hpp

```diff
@@ -1448,298 +1470,110 @@ class test_broker {
-             sid(sid),
-             handle(sco_map.insert_or_assign(topic_filter, *this, 0).first) {}
-
-        ~sub_con_offline() {
-            sco_map.erase(handle, *this);
-        }
-
-        void deliver(
-            MQTT_NS::buffer pub_topic,
-            MQTT_NS::buffer contents,
-            MQTT_NS::qos pub_qos_value,
-            MQTT_NS::retain retain,
-            MQTT_NS::v5::properties props) {
-
-            MQTT_NS::publish_options pubopts = std::min(subopts.get_qos(), pub_qos_value);
-            if (subopts.get_rap() == MQTT_NS::rap::retain && retain == MQTT_NS::retain::yes) {
-                pubopts |= MQTT_NS::retain::yes;
+        void unsubscribe_all() {
+            for (auto const& h : handles) {
+                subs_map_.value().get().erase(h, client_id);
             }
-            if (sid) {
-                props.push_back(MQTT_NS::v5::property::subscription_identifier(*sid));
-            }
-            messages.emplace_back(
-                MQTT_NS::force_move(pub_topic),
-                MQTT_NS::force_move(contents),
-                MQTT_NS::force_move(props),
-                pubopts
-            );
-        }
-
-        bool from_me(endpoint_t const&) const {
-            return false;
+            handles.clear();
         }
 
-        sub_con_offline_map& sco_map;
+        MQTT_NS::optional<std::reference_wrapper<sub_con_map>> subs_map_;
+        con_sp_t con;
         MQTT_NS::buffer client_id;
-        MQTT_NS::buffer topic_filter;
-        std::vector<offline_message> messages;
-        MQTT_NS::subscribe_options subopts;
-        MQTT_NS::optional<std::size_t> sid;
-        sub_con_offline_map::handle handle; // to efficient remove
+
+        MQTT_NS::optional<MQTT_NS::will> will;
+        MQTT_NS::optional<std::chrono::steady_clock::duration> will_delay;
+        MQTT_NS::optional<std::chrono::steady_clock::duration> session_expiry_interval;
+        std::shared_ptr<as::steady_timer> tim_session_expiry;
+        MQTT_NS::optional<MQTT_NS::topic_alias_recv_map_t> topic_alias_recv;
+        // TODO:
+        // insert resend_messages here. get by for_each_store
+        // PUBLISH with assigned packet_id
+        // PUBREL
+        std::deque<offline_message> offline_messages;
+        std::set<sub_con_map::handle> handles; // to efficient remove
     };
-    using mi_sub_con_offline = mi::multi_index_container<
-        sub_con_offline,
+
+    // The mi_session_online container holds the relevant data about an active connection with the broker.
+    // It can be queried either with the clientid, or with the shared pointer to the mqtt endpoint object
+    using mi_session_state = mi::multi_index_container<
+        session_state,
         mi::indexed_by<
-            // Allow multiple client id's for the same topic_filter
+            // non is nullable
             mi::ordered_non_unique<
+                mi::tag<tag_con>,
+                BOOST_MULTI_INDEX_MEMBER(session_state, con_sp_t, con)
```

Comment:
mi_session_state seems to have a lot of indices, they all require a lookup. Are they all required, is it not possible to remove the connection as an index from mi_session_state ?

Reply:
I think it is not a lot. Just 3. All of them are required.

Reply:
But a connection can have a reference to its session, once the session is created. That way, you do not have to lookup the session which belongs to a certain connection. But you have a circular reference then. 

I think a better approach is to attach the session to the protocol handlers instead of the connection.

So instead of (for example):
````C++
ep.set_error_handler(
            [this, wp]
            (MQTT_NS::error_code ec){
````

You would do:
````C++
ep.set_error_handler(
            [this, session]
            (MQTT_NS::error_code ec){
               con_sp_t sp = session.con
````